src/main/java/soot/baf/toolkits/base/OpStackCalculator.java

```diff
@@ -0,0 +1,863 @@
+  }
+
+  private static Stack<Type> updateStack(StackEffectSwitch sw, Stack<Type> st) {
+    @SuppressWarnings("unchecked")
+    Stack<Type> clone = (Stack<Type>) st.clone();
+
+    if (sw.remove_types != null) {
+      if (sw.remove_types.length > clone.size()) {
+        StringBuilder exc = new StringBuilder();
+        exc.append("Expecting values on stack: ");
+        for (Type element : sw.remove_types) {
+          String type = element.toString();
+          if (type.trim().isEmpty()) {
+            type = element instanceof RefLikeType ? "L" : "U";
+          }
+          exc.append(type).append("  ");
+        }
+        exc.append("\n\tbut only found: ");
+        for (Type element : clone) {
+          String type = element.toString();
+          if (type.trim().isEmpty()) {
+            type = element instanceof RefLikeType ? "L" : "U";
+          }
+          exc.append(type).append("  ");
+        }
+
+        if (sw.shouldThrow) {
+          throw new RuntimeException(exc.toString());
+        } else {
+          logger.debug(exc.toString());
+        }
+      }
+      for (int i = sw.remove_types.length - 1; i >= 0; i--) {
+        try {
+          Type t = clone.pop();
+//          if (!checkTypes(t, sw.remove_types[i])) {
+//            System.out.println("Incompatible types: " + t + " : " + sw.remove_types[i]);
+//          }
+        } catch (Exception exc) {
+          return null;
+        }
+      }
+    }
+
+    if (sw.add_types != null) {
+      for (Type element : sw.add_types) {
+        clone.push(element);
+      }
+    }
+
+    return clone;
+  }
+//
+//  private static boolean checkTypes(Type t1, Type t2) {
+//    if (t1 == t2) {
+//      return true;
+//    }
+//    if (t1 instanceof RefLikeType && t2 instanceof RefLikeType) {
+//      return true;
+//    }
+//    if (t1 instanceof IntegerType && t2 instanceof IntegerType) {
+//      return true;
+//    }
+//    if (t1 instanceof LongType && t2 instanceof LongType) {
+//      return true;
+//    }
+//    if (t1 instanceof DoubleType && t2 instanceof DoubleType) {
+//      return true;
+//    }
+//    if (t1 instanceof FloatType && t2 instanceof FloatType) {
+//      return true;
+//    }
+//    return false;
+//  }
+
+  private static RefType isHandlerUnit(Chain<Trap> traps, Unit h) {
+    for (Trap t : traps) {
+      if (t.getHandlerUnit() == h) {
+        return t.getException().getType();
+      }
+    }
+    return null;
+  }
+
+  private static void printStack(StackEffectSwitch sw, PatchingChain<Unit> units, Map<Unit, Stack<Type>> stacks) {
+    try {
+      sw.shouldThrow = false;
+
+      Map<Unit, Integer> indexes = new HashMap<Unit, Integer>();
+      {
+        int count = 0;
+        for (Iterator<Unit> it = units.snapshotIterator(); it.hasNext();) {
+          indexes.put(it.next(), count++);
+        }
+      }
+      for (Iterator<Unit> it = units.snapshotIterator(); it.hasNext();) {
+        Unit unit = it.next();
+        String s = "";
```

Comment:
might be better to use a StringBuilder here

Reply:
Will update.